python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py

```diff
@@ -0,0 +1,536 @@
+        worker_groups_to_clear_delete: Set[NodeType] = field(default_factory=set)
+        # The worker groups that still have workers to be deleted.
+        worker_groups_with_pending_deletes: Set[NodeType] = field(default_factory=set)
+
+    ################################
+    # Interface for ICloudInstanceProvider
+    ################################
+
+    def get_non_terminated(self) -> Dict[CloudInstanceId, CloudInstance]:
+        self._sync_with_api_server()
+        return copy.deepcopy(
+            {
+                id: instance
+                for id, instance in self._cached_instances.items()
+                if instance.status != NodeStatus.TERMINATED
+            }
+        )
+
+    def terminate(self, ids: List[CloudInstanceId], request_id: str) -> None:
+        if request_id in self._requests:
+            # This request is already processed.
+            return
+        self._requests.add(request_id)
+
+        scale_request = self._initialize_scale_request(
+            to_launch={}, to_delete_instances=ids
+        )
+        if scale_request.worker_groups_with_pending_deletes:
+            errors_msg = (
+                "There are workers to be deleted from: "
+                f"{scale_request.worker_groups_with_pending_deletes}. "
+                "Waiting for them to be created before deleting workers."
+            )
+            logger.warning(errors_msg)
+            self._add_terminate_errors(
+                ids,
+                request_id,
+                details=errors_msg,
+            )
+            return
+
+        try:
+            self._submit_scale_request(scale_request)
+        except Exception as e:
+            logger.exception(f"Error terminating nodes: {e}")
+            self._add_terminate_errors(ids, request_id, str(e), e)
+
+    def launch(self, shape: Dict[NodeType, int], request_id: str) -> None:
+        if request_id in self._requests:
+            # This request is already processed.
+            return
+        self._requests.add(request_id)
+
+        scale_request = self._initialize_scale_request(
+            to_launch=shape, to_delete_instances=[]
+        )
+
+        if scale_request.worker_groups_with_pending_deletes:
+            error_msg = (
+                "There are workers to be deleted from: "
+                f"{scale_request.worker_groups_with_pending_deletes}. "
+                "Waiting for them to be deleted before creating new workers."
+            )
+            logger.warning(error_msg)
+            self._add_launch_errors(
+                shape,
+                request_id,
+                details=error_msg,
+            )
+            return
+
+        try:
+            self._submit_scale_request(scale_request)
+        except Exception as e:
+            logger.exception(f"Error launching nodes: {e}")
+            self._add_launch_errors(shape, request_id, str(e), e)
+
+    def poll_errors(self) -> List[CloudInstanceProviderError]:
```

Comment:
Is this only for testing?

Reply:
it's part of the interface for `CloudInstanceProvider`